File: 백준/Gold/1245. 농장 관리/농장 관리.java
import java.awt.*;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {
    private static int N, M;
    private static int[][] arr;
    private static boolean[][] visited;
    private static int[] dx = {-1, 1, 0, 0, -1, -1, 1, 1};
    private static int[] dy = {0, 0, -1, 1, 1, -1, -1, 1};
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        arr = new int[N][M];
        visited = new boolean[N][M];
        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < M; j++) {
                arr[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        int answer = 0;
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (!visited[i][j]) {
                    if (isMountainTop(i, j, arr[i][j])) {
                        answer++;
                    }
                }
            }
        }

        System.out.println(answer);
    }

    private static boolean isMountainTop(int x, int y, int height) {
        Queue<Point> queue = new LinkedList<>();
        //인접한 모든 원소가 높이가 작아야지 산봉우리 가능 (동일 높이 제외)
        visited[x][y] = true;
        queue.add(new Point(x, y));
        boolean isTop = true;
        while (!queue.isEmpty()) {
            Point p = queue.poll();
            for (int i = 0; i < 8; i++) {
                int nx = p.x + dx[i];
                int ny = p.y + dy[i];
                if (nx < 0 || ny < 0 || nx >= N || ny >= M) continue;
                if (arr[nx][ny] > height) isTop = false;
                if (visited[nx][ny]) continue;
                if (arr[nx][ny] == height) {
                    queue.add(new Point(nx, ny));
                    visited[nx][ny] = true;
                }
            }
        }

        if (isTop) {
            return true;
        }

        return false;
    }


}

File: 백준/Gold/1245. 농장 관리/농장 관리.cc
#include <iostream>
#include <queue>
#include <algorithm>

using namespace std;

int arr[100][70];
bool visited[100][70];
int N, M;
int dx[8] = {-1, 1, 0, 0, -1, -1, 1, 1};
int dy[8] = {0, 0, -1, 1, 1, -1, -1, 1};

bool isMountainTop(int x, int y, int height);

int main() {
    cin >> N >> M;
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < M; j++) {
            cin >> arr[i][j];
        }
    }

    int answer = 0;
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < M; j++) {
            if (!visited[i][j]) {
                if (isMountainTop(i, j, arr[i][j])) {
                    answer++;
                }
            }
        }
    }

    cout << answer;
}

bool isMountainTop(int x, int y, int height) {
    queue<pair<int, int>> q;
    q.push(make_pair(x, y));
    visited[x][y] = true;
    bool isTop = true;
    while (!q.empty()) {
        pair<int, int> cur = q.front();
        q.pop();

        for (int i = 0; i < 8; i++) {
            int nx = cur.first + dx[i];
            int ny = cur.second + dy[i];
            if (nx < 0 || ny < 0 || nx >= N || ny >= M) continue;
            if (arr[nx][ny] > height) isTop = false;
            if (visited[nx][ny]) continue;
            if (arr[nx][ny] == height) {
                q.push(make_pair(nx, ny));
                visited[nx][ny] = true;
            }
        }
    }

    if (isTop) {
        return true;
    }
    return false;
};

